Docs 661 add batch api to sql doc (Do not merge yet)

csobier opened the pull request:
For [Docs issue#661](https://github.com/CartoDB/docs/issues/661).  

@dgaubert , I am being proactive and created draft content for the SQL Batch API content.  Let's use this PR as a placeholder for editing and/or new content until this is ready to be "officially released".

csobier commented:
What is the status on when this feature will be "public"?

csobier commented:
@dgaubert (the Travis CI build failed for some reason, not sure why? We'll need to resolve before merging).  
I added the multi query batch info to the Batch API section. Here is the updated staging envn that I generated from the Docs repo. Please review all the Batch API content, and code, and let me know if anything needs to be edited?   http://661-add-batch-api-to-sql-docs.docs.cartodb-staging.com/cartodb-platform/sql-api/sql-batch-api/

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+# SQL Batch API
+
+The SQL Batch API supports a REST endpoint for long-running queries. This is useful for evaluating any timeouts due to endpoint fails. You can use the SQL Batch API to [create](#create-a-job), [read](#read-a-job), [list](#list-jobs), [update](#update-a-job) and [delete](#delete-a-job) queries. You can also run [multiple](#multi-query-batch-jobs) SQL queries in one statement. CartoDB has defined j ...
```
Definitely, I gave you a very bad base doc to make a public documentation, sorry about that. So, I'm going to try to do it in a better way. For introduction, I would use:
```The SQL Batch API allows you to make request with queries that take a considerable time of processing. Usually, that kind of requests raise timeout errors when you use the SQL API. In order to avoid timeouts, you can use the Batch API to create, read, list, update and cancel jobs. Batch API schedules the incoming jobs and allow you to request the results of your jobs.```

This is the idea behind Batch API, so feel free to improve it.

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+# SQL Batch API
+
+The SQL Batch API supports a REST endpoint for long-running queries. This is useful for evaluating any timeouts due to endpoint fails. You can use the SQL Batch API to [create](#create-a-job), [read](#read-a-job), [list](#list-jobs), [update](#update-a-job) and [delete](#delete-a-job) queries. You can also run [multiple](#multi-query-batch-jobs) SQL queries in one statement. CartoDB has defined j ...
+
```
I forgot to add a note to indicate to users that they have to authenticate their request. We can use the same paragraph that we use in `API.md`

> Remember that in order to access ...

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+
+## SQL Batch API Job Schema
+
+The SQL Batch API request to your CartoDB account should include the following elements, that represent the job schema for long-running queries:
```
Probably here, we are giving to the user the idea that he has to fill the job schema in requests, actually everything except `query` is defined by Batch API and are read-only (except `query`). So, I'd say:

`Basically, a job represents SQL statements and includes the following elements: ...`

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+&#124;_ `failed` | job executed but failed, with errors.
+&#124;_ `canceled` | job canceled by user request.
+&#124;_ `unknown` | appears when it is not possible to determine what exactly happened with the job. For example, if draining before exit has failed.
+`query` | the SQL statement to be executed in a database. You can modify the select SQL statement to be used in the job schema.<br/><br/>**Tip:** You can retrieve a query with outputs (SELECT) or inputs (INSERT, UPDATE and DELETE). In some scenarios, you may need to retrieve the query results from a finished job. If that is the case, wrap the query with SELECT * INTO, or CREATE TABLE AS. The resu ...
```
I would remove this sentence: `You can retrieve a query with outputs (SELECT) or inputs (INSERT, UPDATE and DELETE)`

dgaubert commented on doc/sql_batch_api.md:
````diff
@@ -0,0 +1,242 @@
+}
+```
+
+## Delete a Job
````
We should use `Cancel a Job` instead of `Deleate a Job`.

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+
+## Multi Query Batch Jobs
+
+In some cases, you may need to run multiple SQL queries in one statement. The Multi Query batch option enables you run an array of SQL statements, and define the order in which the queries are executed.
```
I'd say:
... you may need to run multiple SQL queries in one **job** ...

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+Name | Description
+--- | ---
+`job_id` | a universally unique identifier (uuid).
+`user_id` | user identifier, as displayed by the username.
```
We have to change `user_id` property by `user`.

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+`user_id` | user identifier, as displayed by the username.
+`status` | displays the result of the long-running query. The possible status results are:
+--- | ---
+&#124;_ `pending` | job waiting to be executed (daily, weekly jobs).
```
`(daily, weekly jobs)` not needed. Please, remove it.

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,242 @@
+&#124;_ `done` | job executed successfully.
+&#124;_ `failed` | job executed but failed, with errors.
+&#124;_ `canceled` | job canceled by user request.
+&#124;_ `unknown` | appears when it is not possible to determine what exactly happened with the job. For example, if draining before exit has failed.
```
`For example, if draining before exit has failed.` it's an implementation detail. it should be removed.

dgaubert commented on doc/sql_batch_api.md:
````diff
@@ -0,0 +1,242 @@
+		“updated_at”: “2015-12-15T07:36:25Z”
+}
+```
+
````
I forgot to explain that user is able to do the same operations as he does with non Multi Query jobs (create, read, list, update and cancel). Can we add some paragraph here to explain that?

csobier replied:
Sure, I just applied all the rest of the edits.  Will go back in and clarify this. Thanks!

dgaubert commented on doc/sql_batch_api.md:
```diff
@@ -0,0 +1,241 @@
+
+- Suppose the first query job status is `"status": "done"`, the second query is `"status": "running"`, and the third query `"status": "pending"`. If the second query fails for some reason, the job status changes to `"status": "failed"` and the last query will not be processed. It is indicated which query failed in the Multi Query batch job
+
+- If you cancel the Multi Query batch job between queries, the job status changes to `"status": "cancelled"` for the Multi Query batch job, but each of the child queries are changed to `"status": "pending"` at the point after it was cancelled. This ensure that no query was cancelled, but the batch array was cancelled
```
I'm not sure about this point, actually I think is quite similar to the first one but with cancel operation. So, I'd try in this way:
If you cancel the Multi Query batch job, the job status changes to `"status": "cancelled"` and if there was a query running then the query changes to `"status": "cancelled"` and further queries of the Multi Query batch job will not be precessed.

What do you think?

dgaubert commented on doc/sql_batch_api.md:
````diff
@@ -0,0 +1,241 @@
+]
+```
+
+If this is the case, make a PATCH request with the following parameters.
````
Whoa! This is outdated, please remove it!

csobier replied:
NO more Patch request?  Okay, removing now!

csobier commented:
@dgaubert , okay- good stuff!  I removed the Patch process and updated the multi query cancel description, as discussed.  Give it another skim and let's make sure it is good to go.

dgaubert commented on doc/sql_batch_api.md:
````diff
@@ -0,0 +1,232 @@
+
+```bash
+HEADERS: PUT /api/v2/sql/job/de305d54-75b4-431b-adb2-eb6b9e546014
+BODY: {
````
Ohh! I provided to you a bad example of this, sorry!! The `BODY` for UPDATE request should be:
```
BODY: {
    "query": “SELECT cartodb_id FROM user_dataset”
}
```

csobier replied:
Ha! It happens.  Editing now....

dgaubert commented on doc/sql_batch_api.md:
````diff
@@ -0,0 +1,232 @@
+	 “updated_at”: “2015-12-17T06:22:42Z”
+}
+```
+
````
Last one thing (I promise), I'd like to add one note here, pretty similar that you added in `Update a job` (line 152):

Note: Jobs can only be cancelled while the status: "pending" or "running", otherwise the SQL Batch API Cancel operation is not allowed. You will receive an error if the job status is anything but “pending” or "running", e.g:

errors: [
	 “The job status is done, cancel is not allowed”
]

Can you add this, please?

csobier commented:
@dgaubert , I added the error code to the Cancel operation. Just need confirmation about the Update operation response code.

dgaubert commented:
:shipit:
